hilti/toolchain/include/ast/node.h

```diff
@@ -172,6 +172,70 @@ using Properties = std::map<std::string, node::PropertyValue>;
 
 } // namespace node
 
+/** Smart pointer wrapping a node, automatically pinning and unpinning it. */
+template<typename T>
+class PinnedNodePtr {
+public:
+    PinnedNodePtr() = default;
+    PinnedNodePtr(T* n) : _node(n) { pin(); }
+    PinnedNodePtr(const PinnedNodePtr& other) : _node(other._node) { pin(); }
+    PinnedNodePtr(PinnedNodePtr&& other) noexcept : _node(other._node) {
+        other._node = nullptr;
+        // reuse the other's pin
+    }
+
+    ~PinnedNodePtr() { unpin(); }
+
+    PinnedNodePtr& operator=(const PinnedNodePtr& other) {
+        if ( this == &other )
+            return *this;
+
+        unpin();
+        _node = other._node;
+        pin();
+        return *this;
+    }
+
+    PinnedNodePtr& operator=(PinnedNodePtr&& other) noexcept {
+        if ( this == &other )
+            return *this;
+
+        unpin();
+        _node = other._node;
+        other._node = nullptr;
+        // reuse the other's pin
+        return *this;
+    }
+
+    void reset() {
+        unpin();
+        _node = nullptr;
+    }
+
+    T* operator->() const { return _node; }
+    T& operator*() const { return *_node; }
+    explicit operator bool() const { return _node != nullptr; }
+    operator T*() const { return _node; }
+    operator Node*() const { return _node; }
```

Comment:
In the usage we already expect `T` to derive from `Node` so this is not needed (and also seems slightly out of place since nothing else in `PinnedNodePtr` makes that requirement explicit).

```suggestion
```

Reply:
Interesting, I had added this because things didn't compile for me without it. But now they do, so removing.